### PR TITLE
[DA-1325] Workbench api

### PR DESCRIPTION
**Pull request by wangy70**

### CODE REVIEW

File: rdr_service/api/workbench_api.py [hunk 1]
```diff
@@ -0,0 +1,23 @@
+from rdr_service import app_util
+
+from rdr_service.api.base_api import BaseApi
+from rdr_service.api_util import WORKBENCH
+from rdr_service.dao.workbench_dao import WorkbenchWorkspaceDao, WorkbenchResearcherDao
+
+
+class WorkbenchWorkspaceApi(BaseApi):
+    def __init__(self):
+        super(WorkbenchWorkspaceApi, self).__init__(WorkbenchWorkspaceDao(), get_returns_children=True)
```
Comment (Seeker1911) [inline rdr_service/api/workbench_api.py]:
```suggestion
        super().__init__(WorkbenchWorkspaceDao(), get_returns_children=True)
```

Comment (Seeker1911) [reply rdr_service/api/workbench_api.py]: same suggestion for remaining super calls.

File: rdr_service/dao/workbench_dao.py [hunk 2]
```diff
@@ -0,0 +1,275 @@
...
+    WorkbenchResearcherHistory,
+    WorkbenchInstitutionalAffiliations,
+    WorkbenchInstitutionalAffiliationsHistory
+)
+from rdr_service.participant_enums import WorkbenchWorkspaceStatus, WorkbenchWorkspaceUserRole
+
+
+class WorkbenchWorkspaceDao(UpdatableDao):
+    def __init__(self):
+        super(WorkbenchWorkspaceDao, self).__init__(WorkbenchWorkspace, order_by_ending=["id"])
```
Comment (Seeker1911) [inline rdr_service/dao/workbench_dao.py]: same

File: rdr_service/dao/workbench_dao.py [hunk 3]
```diff
@@ -0,0 +1,275 @@
...
+            history.workbenchWorkspaceUser = users_history
+            session.add(history)
+
+    def _get_workspace_by_workspace_id_with_session(self, session, workspace_id):
+        return session.query(WorkbenchWorkspace).filter(WorkbenchWorkspace.workspaceSourceId == workspace_id).first()
+
+
+class WorkbenchWorkspaceHistoryDao(UpdatableDao):
+    def __init__(self):
+        super(WorkbenchWorkspaceHistoryDao, self).__init__(WorkbenchWorkspaceHistory, order_by_ending=["id"])
```
Comment (Seeker1911) [inline rdr_service/dao/workbench_dao.py]: same

File: rdr_service/dao/workbench_dao.py [hunk 4]
```diff
@@ -0,0 +1,275 @@
...
+        with self.session() as session:
+            query = session.query(WorkbenchWorkspaceHistory).options(
+                subqueryload(WorkbenchWorkspaceHistory.workbenchWorkspaceUser)
+            )
+            return query.all()
+
+
+class WorkbenchResearcherDao(UpdatableDao):
+    def __init__(self):
+        super(WorkbenchResearcherDao, self).__init__(WorkbenchResearcher, order_by_ending=["id"])
```
Comment (Seeker1911) [inline rdr_service/dao/workbench_dao.py]: same

File: rdr_service/dao/workbench_dao.py [hunk 5]
```diff
@@ -0,0 +1,275 @@
...
+            session.add(history)
+
+    def _get_researcher_by_user_id_with_session(self, session, user_id):
+        return session.query(WorkbenchResearcher).filter(WorkbenchResearcher.userSourceId == user_id)\
+            .order_by(desc(WorkbenchResearcher.created)).first()
+
+
+class WorkbenchResearcherHistoryDao(UpdatableDao):
+    def __init__(self):
+        super(WorkbenchResearcherHistoryDao, self).__init__(WorkbenchResearcherHistory, order_by_ending=["id"])
```
Comment (Seeker1911) [inline rdr_service/dao/workbench_dao.py]: 🙄

File: tests/api_tests/test_workbench_api.py [hunk 6]
```diff
@@ -0,0 +1,335 @@
+from tests.helpers.unittest_base import BaseTestCase
+from rdr_service.dao.workbench_dao import WorkbenchResearcherDao, WorkbenchResearcherHistoryDao, \
+    WorkbenchWorkspaceDao, WorkbenchWorkspaceHistoryDao
+from rdr_service.participant_enums import WorkbenchWorkspaceUserRole
+
+
+class WorkbenchApiTest(BaseTestCase):
+    def setUp(self):
+        super(WorkbenchApiTest, self).setUp(with_data=False)
```
Comment (Seeker1911) [inline tests/api_tests/test_workbench_api.py]: 🤣

File: rdr_service/api/workbench_api.py (hunk 1 above)
Comment (wangy70) [reply rdr_service/api/workbench_api.py]: done

File: rdr_service/dao/workbench_dao.py (hunk 2 above)
Comment (wangy70) [reply rdr_service/dao/workbench_dao.py]: done

File: rdr_service/dao/workbench_dao.py (hunk 3 above)
Comment (wangy70) [reply rdr_service/dao/workbench_dao.py]: done

File: rdr_service/dao/workbench_dao.py (hunk 4 above)
Comment (wangy70) [reply rdr_service/dao/workbench_dao.py]: done

File: rdr_service/dao/workbench_dao.py (hunk 5 above)
Comment (wangy70) [reply rdr_service/dao/workbench_dao.py]: done

File: tests/api_tests/test_workbench_api.py (hunk 6 above)
Comment (wangy70) [reply tests/api_tests/test_workbench_api.py]: done

**Review by Seeker1911 — APPROVED** [review]

**Pull request merged**